### PR TITLE
Global WITH and WITH propagation.

### PR DESCRIPTION
**Pull request by amosbird**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Introduce `enable_global_with_statement` setting  which propagates the first select's `WITH` statements to other select queries at the same level, and makes aliases in `WITH` statements visible to subqueries. 

Detailed description / Documentation draft:

The new setting needs documentation, but it's not clear what description is appropriate.

### CODE REVIEW

File: src/Core/Settings.h [hunk 1]
```diff
@@ -468,6 +468,7 @@ class IColumn;
     M(Bool, output_format_enable_streaming, false, "Enable streaming in output formats that support it.", 0) \
     M(Bool, output_format_write_statistics, true, "Write statistics about read rows, bytes, time elapsed in suitable output formats.", 0) \
     M(Bool, allow_non_metadata_alters, true, "Allow to execute alters which affects not only tables metadata, but also data on disk", 0) \
+    M(Bool, with_global, false, "Enable WITH statements for UNION queries", 0) \
```
Comment (KochetovNicolai) [inline src/Core/Settings.h:471]: `global_with` sounds better
`allow_global_with` or `enable_global_with` even better

Comment (alexey-milovidov) [reply src/Core/Settings.h:471]: We don't have `enable` neither `allow` for every setting that enables or allows something...

Comment (KochetovNicolai) [reply src/Core/Settings.h:471]: It is good to have a verb in boolean setting name. e.g. `output_format_write_statistics`

**Comment by KochetovNicolai** [comment]
```
set with_global = 1;
with 10 as x select x + y from (select 2 as y) union all select x + 1 from (select 1 as x);

WITH 10 AS x
SELECT x + y
FROM 
(
    SELECT 2 AS y
)
UNION ALL
SELECT x + 1
FROM 
(
    SELECT 1 AS x
)


Received exception from server (version 20.10.1):
Code: 179. DB::Exception: Received from localhost:9000. DB::Exception: Different expressions with the same alias x:
1 AS x
and
10 AS x
```

Works when `with_global = 0`

**Comment by KochetovNicolai** [comment]
```
set with_global = 1;
explain syntax with 1 as x select 1 + y from (select x + 2 as y) union all select x + y from (select 1 as y);

EXPLAIN SYNTAX
WITH 1 AS x
SELECT 1 + y
FROM 
(
    SELECT x + 2 AS y
)
UNION ALL
SELECT x + y
FROM 
(
    SELECT 1 AS y
)


Received exception from server (version 20.10.1):
Code: 47. DB::Exception: Received from localhost:9000. DB::Exception: Missing columns: 'x' while processing query: 'SELECT x + y FROM (SELECT 1 AS y)', required columns: 'x' 'y', source columns: 'y'. 

```
That's unusual.  Query without explain works.

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 2]
```diff
@@ -250,6 +251,8 @@ InterpreterSelectQuery::InterpreterSelectQuery(
         source_header = input_pipe->getHeader();
     }
 
+    if (context->getSettingsRef().with_global)
+        ApplyWithAliasVisitor().visit(query_ptr);
```
Comment (KochetovNicolai) [inline src/Interpreters/InterpreterSelectQuery.cpp:255]: I think it need to be moved into `TreeRewriter`. Probably,  `TreeRewriter::normalize`

File: src/Interpreters/InterpreterSelectWithUnionQuery.cpp [hunk 3]
```diff
@@ -38,6 +39,9 @@ InterpreterSelectWithUnionQuery::InterpreterSelectWithUnionQuery(
     if (!num_selects)
         throw Exception("Logical error: no children in ASTSelectWithUnionQuery", ErrorCodes::LOGICAL_ERROR);
 
+    if (context->getSettingsRef().with_global)
+        ApplyWithGlobalVisitor().visit(query_ptr);
```
Comment (KochetovNicolai) [inline src/Interpreters/InterpreterSelectWithUnionQuery.cpp:43]: And this too.

**Review by KochetovNicolai — CHANGES_REQUESTED** [review]
Generally it is fine.

Probably, it would be slightly better to implement this feature without changing of AST. We will need to think about it before CTE implementation.

File: src/Interpreters/InterpreterSelectWithUnionQuery.cpp (hunk 3 above)
Comment (amosbird) [reply src/Interpreters/InterpreterSelectWithUnionQuery.cpp]: done

File: src/Interpreters/InterpreterSelectQuery.cpp (hunk 2 above)
Comment (amosbird) [reply src/Interpreters/InterpreterSelectQuery.cpp]: It's somehow consistent with the current CTE implementation ...

**Review by KochetovNicolai — APPROVED** [review]

**Pull request merged**